TM-14 Task Details Component

TomsBemis opened the pull request:
**1) Edit button to switch to edit mode for updating selected task:**
<img width="740" alt="TM-14 PR 1" src="https://github.com/TomsBemis/task-manager-frontend/assets/149372128/1324e898-6f38-4376-93af-0e96686d6501">
**2) Edit task form where new values are set:**
<img width="744" alt="TM-14 PR 2" src="https://github.com/TomsBemis/task-manager-frontend/assets/149372128/628fe5ce-bfc2-484e-8a7c-6f33f3957b06">
**3) Task values changed after submitting the form:**
<img width="737" alt="TM-14 PR 3" src="https://github.com/TomsBemis/task-manager-frontend/assets/149372128/a1758b22-11f9-44d8-be5e-3d18e2f19f0d">
**4) Custom validator for task title that ignores the task being edited when checking if title is unique:**
<img width="738" alt="TM-14 PR 4" src="https://github.com/TomsBemis/task-manager-frontend/assets/149372128/2313672a-cacb-4892-bb1d-85c2db9f10d7">

TomsBemis committed:
TM-14 Task can be updated from the task details component

TomsBemis committed:
TM-10 Incorrect task id on update fix

wanpii commented on src/app/tasks/task-details/task-details.component.html:
```diff
@@ -1,22 +1,92 @@
-                <button class="btn btn-danger" 
-                (click)="this.task !== null && this.task.id !== null ? onDeleted(this.task.id) : true">
-                Delete</button>
+    <form [formGroup]="editTaskForm!" (ngSubmit)="onSubmit()">
```
editTaskForm! ?

wanpii commented on src/app/tasks/task-details/task-details.component.html:
```diff
@@ -1,22 +1,92 @@
+                </div>
+            </div>
+        </div>    
+        @if (this.editMode == true) {
```
you can just leave it as @if(this.editMode) means true !this.editMode would mean false

wanpii commented on src/app/tasks/task-details/task-details.component.ts:
```diff
@@ -1,28 +1,111 @@
   ngOnInit(): void {
-    this.task = this.taskService.getTask(this.route.snapshot.params['id']);
+    
+    this.task = this.taskService.getTask(this.route.snapshot.params['id']) ?? emptyTask;
```
to have 100% certainty here, you can subscribe to the activatedROute params, get id param and then get the task, that way you won't have to give it a default value of emptyTask

wanpii commented on src/app/tasks/task-details/task-details.component.ts:
```diff
@@ -1,28 +1,111 @@
+          type: this.taskTypes.find( taskType => 
+            taskType.value == this.editTaskForm.get('type')?.value
+          ) ?? null,
+          createdOn: new Date(),
```
createdOn shouldn't be changed in my opinion, since you don't create a new task, but update it

wanpii commented on src/app/tasks/task-details/task-details.component.ts:
```diff
@@ -1,28 +1,111 @@
+      status: new FormControl(this.task?.status?.value, Validators.required)
+    });
+
+    console.log(this.editTaskForm);
```
remove console.log

wanpii commented on src/app/tasks/task.service.ts:
```diff
@@ -51,6 +51,19 @@ export class TaskService {
+        
+        this.taskList.tasks.splice(
+            this.taskList.tasks
+            .findIndex(task => task.id === id ), 1);
```
you can instead of splicing, just update the existing entry with .map()

TomsBemis committed:
TM-14 Pull Request 5 Requested changes 1

Attribute createdOn is not changed when updating task. Added new attribute modifiedOn to hold a timestamp of last update. Task details template changed to not show anything if task is null (invalid task id in url parameters). Task details component subscribes to route parameters to fetch task id. Task Service updates task using a map not splicing and pushing.

wanpii commented on src/app/tasks/create-task/create-task.component.ts:
```diff
@@ -43,6 +42,7 @@ export class CreateTaskComponent {
       type: this.taskTypes.find( taskType => 
         taskType.value == this.createTaskForm.get('type')?.value
       ) ?? null,
+      modiefiedOn: new Date(),
```
typo

wanpii commented on src/app/tasks/task-details/task-details.component.ts:
```diff
@@ -1,28 +1,116 @@
+  taskTypes : Option[] = this.taskService.getTaskTypes();
+  taskStatuses : Option[] = this.taskService.getTaskStatuses();
+  routeParamsSubscription: Subscription = this.route.params.subscribe((params: Params) => {
+    this.task = this.taskService.getTask(+params['id']);
```
whats the + for?

wanpii commented on src/app/tasks/task.model.ts:
```diff
@@ -10,6 +10,7 @@ export interface Task {
     description: string, 
     type: Option | null, 
     createdOn: Date, 
+    modiefiedOn: Date, 
```
typo

wanpii commented on src/app/tasks/task.model.ts:
```diff
@@ -19,5 +20,6 @@ export const emptyTask: Task = {
     description: '',
     type: null,
     createdOn: new Date(),
+    modiefiedOn: new Date(),
```
typo

wanpii commented on src/app/tasks/task.service.ts:
```diff
@@ -29,6 +30,7 @@ export class TaskService {
                 description: taskElementData.description,
                 type: this.taskTypes.find(taskType => taskType.value == taskElementData.type) ?? null,
                 createdOn: new Date(taskElementData.createdOn),
+                modiefiedOn: new Date(taskElementData.modifiedOn),
```
typo

wanpii commented on src/app/tasks/task.service.ts:
```diff
@@ -51,6 +53,18 @@ export class TaskService {
+            return task
+        })
+
+        return editedTask;
```
are you using this returned task anywhere?

wanpii commented on src/app/tasks/task-details/task-details.component.ts:
```diff
@@ -1,28 +1,116 @@
 
-  constructor(private router: Router, private route: ActivatedRoute, private taskService: TaskService) {}
+  onEdit() {
+    this.editMode = true;
```
could be nice to maybe just give it the opposite value? (everything below is optional)
this.editMode = !this.editMode?
That way you can toggle between editMode, though if you decide to implement this then you have to reset the form to the initial values whenever you close the editMode

TomsBemis committed:
TM-14 Pull Request 5 Requested changes 2

TomsBemis committed:
TM-14 Pull Request 5 Requested changes 2

Removed updateTask returning a task because it is only used in one place. Task Details component edit button toggles edit mode.
